File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/textures/ColourTexture.java
package org.saar.lwjgl.opengl.textures;

import org.lwjgl.system.MemoryUtil;
import org.saar.lwjgl.opengl.utils.MemoryUtils;

import java.nio.ByteBuffer;

public class ColourTexture implements ITexture {

    private final ITexture texture;
    private final int r;
    private final int g;
    private final int b;
    private final int a;

    public ColourTexture(ITexture texture, int r, int g, int b, int a) {
        this.texture = texture;
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    public static ColourTexture of(int r, int g, int b, int a) {
        final Texture2D texture = new Texture2D(1, 1);
        final ByteBuffer buffer = MemoryUtils.allocByte(4);
        buffer.put((byte) r).put((byte) g).put((byte) b).put((byte) a).flip();
        texture.load(buffer);
        MemoryUtil.memFree(buffer);
        return new ColourTexture(texture, r, g, b, a);
    }

    public int getRed() {
        return this.r;
    }

    public int getGreen() {
        return this.g;
    }

    public int getBlue() {
        return this.b;
    }

    public int getAlpha() {
        return this.a;
    }

    private ITexture getTexture() {
        return this.texture;
    }

    @Override
    public void bind(int unit) {
        getTexture().bind(unit);
    }

    @Override
    public void bind() {
        getTexture().bind();
    }

    @Override
    public void unbind() {
        getTexture().unbind();
    }

    @Override
    public void delete() {
        getTexture().delete();
    }
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/texture/ColourTexture.kt
package org.saar.lwjgl.opengl.texture

import org.saar.lwjgl.opengl.constants.DataType
import org.saar.lwjgl.opengl.constants.FormatType
import org.saar.lwjgl.opengl.constants.InternalFormat
import org.saar.lwjgl.util.buffer.LwjglByteBuffer

class ColourTexture private constructor(
    private val texture: ReadOnlyTexture,
    val red: Int,
    val green: Int,
    val blue: Int,
    val alpha: Int
) : ReadOnlyTexture2D {

    override fun bind(unit: Int) = this.texture.bind(unit)

    override fun bind() = this.texture.bind()

    override fun unbind() = this.texture.unbind()

    override fun delete() = this.texture.delete()

    override fun getWidth(): Int = 1

    override fun getHeight(): Int = 1

    companion object {
        @JvmStatic
        fun of(r: Int, g: Int, b: Int, a: Int): ColourTexture {
            LwjglByteBuffer.allocate(4).use { buffer ->
                buffer.put(r.toByte()).put(g.toByte()).put(b.toByte()).put(a.toByte()).flip()
                val texture = Texture2D.of(1, 1, InternalFormat.RGBA2, 1).apply {
                    load(0, FormatType.RGBA, DataType.U_BYTE, buffer.asByteBuffer())
                }
                return ColourTexture(texture, r, g, b, a)
            }
        }
    }
}